### ISSUE TITLE
Created a JSON-based API.

### ISSUE THREAD

**Issue by ExpDev07** [CSSEGISandData/COVID-19#8]
Hello and thank you so much for maintaining this! I have created an API that reads your data and returns it in a way that's more friendly to use in programs. It also supports history. It's still a W.I.P and I would love contributions! It is open-sourced here: https://github.com/ExpDev07/coronavirus-tracker-api. Feel free to use it in your projects!

It is very fast due to caching.

**The current endpoints are (more will be added):**
https://coronavirus-tracker-api.herokuapp.com/confirmed
https://coronavirus-tracker-api.herokuapp.com/deaths
https://coronavirus-tracker-api.herokuapp.com/recovered

For all of them combined:
https://coronavirus-tracker-api.herokuapp.com/all

**Comment by brent-field** [comment]
This is super helpful and works swimmingly. Thanks!

**Comment by altezza04** [comment]
Thank you! I start to use it on my site.

**Comment by ExpDev07** [comment]
Glad to hear you're using it! @altezza04 . Feel free to share your site :)! I'd love to see it.

**Comment by bjarkimg** [comment]
WIP - Power BI
http://bit.ly/38qeMbz

**Comment by isratahm** [comment]
Hi- thank you so much for this! (Context I'm very much a beginner at this, but I thought I'd give it a go)

And I seemed to have followed all the instructions and have gotten everything set up. But since it's my first time I must have made some sort of error because I keep getting a 404.

Here is the exact error "GET / HTTP/1.1" 404 -

Would really appreciate it if you could help! Thanks (:

**Comment by bjarkimg** [comment]
@ExpDev07 possible to return the date fields unpivoted ? 
Can do this in power bi just a bit of hassle.

grate job btw :)

**Comment by 3verse** [comment]
> WIP - Power BI
> http://bit.ly/38qeMbz
hi @bjarkimg - nice viz! I created my own on Tableau public (https://tabsoft.co/2OKTU72) but this was before the dataset was moved to github, I'll have work on the datasource again to update this now

**Comment by bjarkimg** [comment]
@3verse nice viz :), yea did the google sheet first.. hope this will be stable.

**Comment by altezza04** [comment]
> Glad to hear you're using it! @altezza04 . Feel free to share your site :)! I'd love to see it.

http://chinavirus.live/
I use it on the world ranking part only, it seem the data is not updated often , but anyway, thanks for the effort!

**Comment by bjarkimg** [comment]
FYI added Deaths and Recovered, with Measure select option. 
http://bit.ly/38qeMbz

**Comment by bjarkimg** [comment]
> @ExpDev07 possible to return the date fields unpivoted ?
> Can do this in power bi just a bit of hassle.
> 
> grate job btw :)

@ExpDev07 Ignore this request, solved in Power BI, auto update should work now. thanks for the API!

**Comment by chovy** [comment]
Looks to be broken now.

**Comment by ExpDev07** [comment]
@chovy it broke when they re-formatted the repo, I've fixed it to point to the correct paths now and it's working again.

**Comment by chovy** [comment]
awesome. thank you sir.

**Comment by bjarkimg** [comment]
FYI
Added Forecast on Page 3
http://bit.ly/38qeMbz

![image](https://user-images.githubusercontent.com/1289419/74649777-27919880-5178-11ea-9fe0-4cd295030ad6.png)

**Comment by sorxrob** [comment]
I made a website too https://the2019ncov.com

**Comment by leopoul** [comment]
![Screenshot 2020-02-24 at 21 25 59](https://user-images.githubusercontent.com/770064/75188655-cd32a200-574c-11ea-84b1-50eaddd49f38.png)

I parse your API endpoint, massage it and throw it in an influxDB. At the same time I build a unique list of locations to feed the Grafana WorldMap plugin served via Flask. Here is a first draft of the Grafana dashboard. I plan to work a bit more around my code and share dashboard, parser and Flask server code.

**Comment by iceweasel1** [comment]
@leopoul Grafana and InfluxDB, I love it. Nice work!

**Comment by smartcuc** [comment]
Cool, how did you get the data into influxDB?

**Comment by chovy** [comment]
I noticed a map that has quanrtined data. Can we add that?

**Comment by ExpDev07** [comment]
@chovy ill see what I can do!:)

**Comment by leopoul** [comment]
Code is here: https://github.com/leopoul/covid19_grafana . It's a quick and dirty release so feel free to contribute. I transform the data to get it into grafana and extract the unique locations to make it work with the Grafana plugin. I wanted to have something quickly out so there is lots of room for improvement.
![grafana_1](https://user-images.githubusercontent.com/770064/75441935-d1360e00-595e-11ea-856d-a61a10d44474.png)
![grafana_2](https://user-images.githubusercontent.com/770064/75441966-d6935880-595e-11ea-9ef2-51f8396b264d.png)

**Comment by elmortero** [comment]
Great work! Is there a possibility to filter by country in the request?

**Comment by ExpDev07** [comment]
Please head over to https://github.com/ExpDev07/coronavirus-tracker-api for any questions/discussions regarding the API. :)!

**Comment by Laeyoung** [comment]
Oh nice work! 

I did similar thing : )!
https://github.com/Laeyoung/COVID-19-API

**Comment by mauro-gobbi** [comment]
Hello, this API is never up to date and not in sync with the daily reports

**Comment by ExpDev07** [comment]
Hello @mauro-gobbi. It caches the daily reports for 6 hours. So at most, it takes 6 hours for it to show the changes after an update. This is to allow for faster results and avoid having to make calls to GitHub per request. If you fork the API, you can change the cache duration to an hour, 30 minutes, or completely remove it if you wish. It's as simply as just changing a variable.

**Comment by mauro-gobbi** [comment]
> Hello @mauro-gobbi. It caches the daily reports for 6 hours. So at most, it takes 6 hours for it to show the changes after an update. This is to allow for faster results and avoid having to make calls to GitHub per request. If you fork the API, you can change the cache duration to an hour, 30 minutes, or completely remove it if you wish. It's as simply as just changing a variable. Feel also free to open an issue or pull request if you believe it should be lowered. I haven't had any complaints besides yours, so most seem OK with the delay.
> **EDIT:** It’s cached for 1 hour.

Hi, thanks a lot for your reply and thanks for making this API available. 😊
Sorry but I do not understand, even if you refresh the data every hour, and regardless caching results, isn't the data source (the .csv in "csse_covid_19_time_series") updated on once a day from J. Hopkins?
It's probably my fault but I do not actually understand how to get fresh data more frequently than one time per day.

**Comment by ExpDev07** [comment]
> isn't the data source (the .csv in "csse_covid_19_time_series") updated only once a day from J. Hopkins
Yes, you are correct in that. So whenever JSU updates their time series, it takes a maximum of 1 hour and a minimum of 0 seconds for it to reflect in the API (which is why the data may differ). If you're talking about live updates, that's not really a thing. Most sources will give you daily updates.

**Comment by mauro-gobbi** [comment]
> > isn't the data source (the .csv in "csse_covid_19_time_series") updated only once a day from J. Hopkins
> 
> Yes, you are correct in that. So whenever JSU updates their time series, it takes a maximum of 1 hour and a minimum of 0 seconds for it to reflect in the API (which is why the data may differ). If you're talking about live updates, that's not really a thing. Most sources will give you daily updates.

Ok great, fair enough. Live updates are not possible for sure, I hope they will start to upload a new .csv report at least every 12h. Thanks again 👍

**Comment by Omaroid** [comment]
Hello everyone,

I maid an android app calling @ExpDev07 Rest API, check the repo. https://github.com/Omaroid/Covid-19-Android

**Comment by MohammedAlrozzi** [comment]
@ExpDev07 thanks for the great work. could you please explain on how to get confirmed cases for example for one single country. how the GET url will look like. Thanks a lot.

**Comment by Jeff-Lewis** [comment]
Bjarki @bjarkimg Can you share your Power BI pbix file please? Or is it possible to make the PBI clonable/save as on app.powerbi.com?

**Comment by chovy** [comment]
api missing data again.

**Comment by roenw** [comment]
@ExpDev07 Your API works great. I have one hosted locally since I’m developing my own [coronavirus dashboard](https://covid19.roen.us).

Typically how often do the numbers presented update? Thank you.

**Comment by DevakiTrishani** [comment]
@ExpDev07 : I'm new to Power BI, and would like to use the data in this Repo. The problem is I don't know how to connect said data in Power BI.
Please can you help. Sorry for asking on this thread, I have no other means to ask you.
btw, your Dasshboards look impressive.
Thank you.

**Comment by DevakiTrishani** [comment]
@bjarkimg : I'm new to Power BI, and would like to use the data in this Repo. The problem is I don't know how to connect said data in Power BI.
Please can you help. Sorry for asking on this thread, I have no other means to ask you.
btw, your Dashboards look impressive.
Thank you.

**Comment by ExpDev07** [comment]
@roenw Updates whenever this repo updates (with a delay of max an hour for cache to reload) :).

**Comment by roenw** [comment]
Thanks @ExpDev07!

**Comment by Aurel-PBI** [comment]
> FYI
> Added Forecast on Page 3
> http://bit.ly/38qeMbz
> 
> ![image](https://user-images.githubusercontent.com/1289419/74649777-27919880-5178-11ea-9fe0-4cd295030ad6.png)

Hi, bjarkimg 

Could you share your pbix file. I would like to learn how you do this amazing dashboards
Thanks again

**Comment by Aurel-PBI** [comment]
> > @ExpDev07 possible to return the date fields unpivoted ?
> > Can do this in power bi just a bit of hassle.
> > grate job btw :)
> 
> @ExpDev07 Ignore this request, solved in Power BI, auto update should work now. thanks for the API!

Hi bjarkimg 
How did u unpivoted the date with PowerBi with data Json AND with an auto update.
I'm realy interested to know.
Thanks again

**Comment by DevakiTrishani** [comment]
@ExpDev07 Thank You!

**Comment by itsgorain** [comment]
Are individual US states data no longer coming through in recoveries and deaths?

**Comment by ExpDev07** [comment]
@itsgorain JHU has dropped tracking of recoveries completely, for confirmed and deaths, they have decided to do U.S.A as a whole instead of individual states. These changes obviously are reflected in the API.

**Comment by ajmishra1989** [comment]
I'm a beginer here and wanted to store how we can store this data in Google cloud storage first (Using the API). Would be helpful if anyone can help me with the exact steps. 

Much thanks in advance.

**Comment by ExpDev07** [comment]
@ajmishra1989 I'm guessing you'd have to write a script to periodically retrieve the data you want and push it to Google Cloud Storage.

**Comment by ajmishra1989** [comment]
> @ajmishra1989 I'm guessing you'd have to write a script to periodically retrieve the data you want and push it to Google Cloud Storage.

OK, Thanks. Any possible workaround my second question? I'm looking for place to fetch data for user sentiments around Covid-19. 

Thanks.

**Comment by ExpDev07** [comment]
I believe there’s some scrapers out there that offers a news feed etc, google around “Covid19 news scraper github” @ajmishra1989

**Comment by sorxrob** [comment]
Good bye recovered data.

**Comment by Omaroid** [comment]
> Good bye recovered data.

They added the recovered data !

**Comment by mahdighazali** [comment]
> > > @ExpDev07 possible to return the date fields unpivoted ?
> > > Can do this in power bi just a bit of hassle.
> > > grate job btw :)
> > 
> > 
> > @ExpDev07 Ignore this request, solved in Power BI, auto update should work now. thanks for the API!
> 
> Hi bjarkimg
> How did u unpivoted the date with PowerBi with data Json AND with an auto update.
> I'm realy interested to know.
> Thanks again

Source = Json.Document(Web.Contents("https://coronavirus-tracker-api.herokuapp.com/all")),
    #"Converted to Table" = Record.ToTable(Source),
    #"Expanded Value" = Table.ExpandRecordColumn(#"Converted to Table", "Value", {"last_updated", "latest", "locations", "source"}, {"Value.last_updated", "Value.latest", "Value.locations", "Value.source"}),
    #"Expanded Value.locations" = Table.ExpandListColumn(#"Expanded Value", "Value.locations"),
    #"Expanded Value.locations1" = Table.ExpandRecordColumn(#"Expanded Value.locations", "Value.locations", {"coordinates", "country", "country_code", "history", "latest", "province"}, {"Value.locations.coordinates", "Value.locations.country", "Value.locations.country_code", "Value.locations.history", "Value.locations.latest", "Value.locations.province"}),
    #"Expanded Value.locations.coordinates" = Table.ExpandRecordColumn(#"Expanded Value.locations1", "Value.locations.coordinates", {"lat", "long"}, {"Value.locations.coordinates.lat", "Value.locations.coordinates.long"}),
    #"Expanded Value.locations.history" = Table.ExpandRecordColumn(#"Expanded Value.locations.coordinates", "Value.locations.history", {"1/22/20", "1/23/20", "1/24/20", "1/25/20", "1/26/20", "1/27/20", "1/28/20", "1/29/20", "1/30/20", "1/31/20", "2/1/20", "2/10/20", "2/11/20", "2/12/20", "2/13/20", "2/14/20", "2/15/20", "2/16/20", "2/17/20", "2/18/20", "2/19/20", "2/2/20", "2/20/20", "2/21/20", "2/22/20", "2/23/20", "2/24/20", "2/25/20", "2/26/20", "2/27/20", "2/28/20", "2/29/20", "2/3/20", "2/4/20", "2/5/20", "2/6/20", "2/7/20", "2/8/20", "2/9/20", "3/1/20", "3/10/20", "3/11/20", "3/12/20", "3/13/20", "3/14/20", "3/15/20", "3/16/20", "3/17/20", "3/18/20", "3/19/20", "3/2/20", "3/20/20", "3/21/20", "3/22/20", "3/23/20", "3/24/20", "3/25/20", "3/26/20", "3/27/20", "3/3/20", "3/4/20", "3/5/20", "3/6/20", "3/7/20", "3/8/20", "3/9/20"}, {"Value.locations.history.1/22/20", "Value.locations.history.1/23/20", "Value.locations.history.1/24/20", "Value.locations.history.1/25/20", "Value.locations.history.1/26/20", "Value.locations.history.1/27/20", "Value.locations.history.1/28/20", "Value.locations.history.1/29/20", "Value.locations.history.1/30/20", "Value.locations.history.1/31/20", "Value.locations.history.2/1/20", "Value.locations.history.2/10/20", "Value.locations.history.2/11/20", "Value.locations.history.2/12/20", "Value.locations.history.2/13/20", "Value.locations.history.2/14/20", "Value.locations.history.2/15/20", "Value.locations.history.2/16/20", "Value.locations.history.2/17/20", "Value.locations.history.2/18/20", "Value.locations.history.2/19/20", "Value.locations.history.2/2/20", "Value.locations.history.2/20/20", "Value.locations.history.2/21/20", "Value.locations.history.2/22/20", "Value.locations.history.2/23/20", "Value.locations.history.2/24/20", "Value.locations.history.2/25/20", "Value.locations.history.2/26/20", "Value.locations.history.2/27/20", "Value.locations.history.2/28/20", "Value.locations.history.2/29/20", "Value.locations.history.2/3/20", "Value.locations.history.2/4/20", "Value.locations.history.2/5/20", "Value.locations.history.2/6/20", "Value.locations.history.2/7/20", "Value.locations.history.2/8/20", "Value.locations.history.2/9/20", "Value.locations.history.3/1/20", "Value.locations.history.3/10/20", "Value.locations.history.3/11/20", "Value.locations.history.3/12/20", "Value.locations.history.3/13/20", "Value.locations.history.3/14/20", "Value.locations.history.3/15/20", "Value.locations.history.3/16/20", "Value.locations.history.3/17/20", "Value.locations.history.3/18/20", "Value.locations.history.3/19/20", "Value.locations.history.3/2/20", "Value.locations.history.3/20/20", "Value.locations.history.3/21/20", "Value.locations.history.3/22/20", "Value.locations.history.3/23/20", "Value.locations.history.3/24/20", "Value.locations.history.3/25/20", "Value.locations.history.3/26/20", "Value.locations.history.3/27/20", "Value.locations.history.3/3/20", "Value.locations.history.3/4/20", "Value.locations.history.3/5/20", "Value.locations.history.3/6/20", "Value.locations.history.3/7/20", "Value.locations.history.3/8/20", "Value.locations.history.3/9/20"}),
    #"Removed Columns" = Table.RemoveColumns(#"Expanded Value.locations.history",{"Value.source"}),
    #"Reordered Columns" = Table.ReorderColumns(#"Removed Columns",{"Name", "Value.last_updated", "Value.latest", "Value.locations.latest", "Value.locations.province", "Value.locations.coordinates.lat", "Value.locations.coordinates.long", "Value.locations.country", "Value.locations.country_code", "Value.locations.history.1/22/20", "Value.locations.history.1/23/20", "Value.locations.history.1/24/20", "Value.locations.history.1/25/20", "Value.locations.history.1/26/20", "Value.locations.history.1/27/20", "Value.locations.history.1/28/20", "Value.locations.history.1/29/20", "Value.locations.history.1/30/20", "Value.locations.history.1/31/20", "Value.locations.history.2/1/20", "Value.locations.history.2/10/20", "Value.locations.history.2/11/20", "Value.locations.history.2/12/20", "Value.locations.history.2/13/20", "Value.locations.history.2/14/20", "Value.locations.history.2/15/20", "Value.locations.history.2/16/20", "Value.locations.history.2/17/20", "Value.locations.history.2/18/20", "Value.locations.history.2/19/20", "Value.locations.history.2/2/20", "Value.locations.history.2/20/20", "Value.locations.history.2/21/20", "Value.locations.history.2/22/20", "Value.locations.history.2/23/20", "Value.locations.history.2/24/20", "Value.locations.history.2/25/20", "Value.locations.history.2/26/20", "Value.locations.history.2/27/20", "Value.locations.history.2/28/20", "Value.locations.history.2/29/20", "Value.locations.history.2/3/20", "Value.locations.history.2/4/20", "Value.locations.history.2/5/20", "Value.locations.history.2/6/20", "Value.locations.history.2/7/20", "Value.locations.history.2/8/20", "Value.locations.history.2/9/20", "Value.locations.history.3/1/20", "Value.locations.history.3/10/20", "Value.locations.history.3/11/20", "Value.locations.history.3/12/20", "Value.locations.history.3/13/20", "Value.locations.history.3/14/20", "Value.locations.history.3/15/20", "Value.locations.history.3/16/20", "Value.locations.history.3/17/20", "Value.locations.history.3/18/20", "Value.locations.history.3/19/20", "Value.locations.history.3/2/20", "Value.locations.history.3/20/20", "Value.locations.history.3/21/20", "Value.locations.history.3/22/20", "Value.locations.history.3/23/20", "Value.locations.history.3/24/20", "Value.locations.history.3/25/20", "Value.locations.history.3/26/20", "Value.locations.history.3/27/20", "Value.locations.history.3/3/20", "Value.locations.history.3/4/20", "Value.locations.history.3/5/20", "Value.locations.history.3/6/20", "Value.locations.history.3/7/20", "Value.locations.history.3/8/20", "Value.locations.history.3/9/20"}),
     Cols = Table.ColumnNames(#"Reordered Columns"),
    ColsUnPivot = List.Skip(Cols,9),
    Unpivot = Table.Unpivot( #"Reordered Columns",ColsUnPivot,"Attribute","Value"),

**Comment by mahdighazali** [comment]
Hi,
Basically:
Move the data columns (before expand) to be the last one,
Create a variable to get all the column names, ( Cols = Table.ColumnNames(#"Reordered Columns"),)

Create a variable all the columns (by excluding the one who are not part of the unpivot) ColsUnPivot = List.Skip(Cols,9),

Unpivot the columns:
Unpivot = Table.Unpivot( #"Reordered Columns",ColsUnPivot,"Attribute","Value"),

**Comment by mahdighazali** [comment]
> Hello and thank you so much for maintaining this! I have created an API that reads your data and returns it in a way that's more friendly to use in programs. It also supports history. It's still a W.I.P and I would love contributions! It is open-sourced here: https://github.com/ExpDev07/coronavirus-tracker-api. Feel free to use it in your projects!
> 
> It is very fast due to caching.
> 
> **The current endpoints are (more will be added):**
> https://coronavirus-tracker-api.herokuapp.com/confirmed
> https://coronavirus-tracker-api.herokuapp.com/deaths
> https://coronavirus-tracker-api.herokuapp.com/recovered
> 
> For all of them combined:
> https://coronavirus-tracker-api.herokuapp.com/all

Hi @ExpDev07 ,
Thank you for the API. It's really helpfull...Please if you can provide me when you refresh the data...It's 2020-03-29 and the last refresh of the data is 2020-03-27 :)
Thank you very much for your help :)

**Comment by rolson21** [comment]
Hey All, I am trying to set up a Power BI map for US Counties, but literally have no clue what I am doing. Can anyone give me a quick tutorial to connect this? I was able to get a Shape Map with US Counties, but how to connect the confirmed coronavirus cases to it??

**Comment by joefrey** [comment]
@ExpDev07 Is there any way to update the json files same as the worldometer. I checked the json [https://coronavirus-tracker-api.herokuapp.com/confirmed](https://coronavirus-tracker-api.herokuapp.com/confirmed) but it's a bit delayed.

**Comment by RafiqCR7** [comment]
@ExpDev07 appreciate this but it doesn't look like there is a report available for the 8th yet, however the api is reporting the last updated date on the figures provided as the 8th? Surely it should be reporting these with the last updated date set to the 7th as this is when the figures are from?

**Comment by zamponotiropita** [comment]
The endpoint https://coronavirus-tracker-api.herokuapp.com/all seems to be down at the moment (Status 503 Service Unavailable)
![image](https://user-images.githubusercontent.com/23146389/79633009-0989e980-8163-11ea-845d-ef69717fe2ec.png)

**Comment by CSSEGISandData** [comment]
Hello! We are also providing JSON query of the feature layer, such as this [one](https://services1.arcgis.com/0MSEUqKaxRlEPj5g/arcgis/rest/services/ncov_cases/FeatureServer/1/query?where=1%3D1&objectIds=&time=&geometry=&geometryType=esriGeometryEnvelope&inSR=&spatialRel=esriSpatialRelIntersects&resultType=none&distance=0.0&units=esriSRUnit_Meter&returnGeodetic=false&outFields=*&returnGeometry=true&featureEncoding=esriDefault&multipatchOption=xyFootprint&maxAllowableOffset=&geometryPrecision=&outSR=&datumTransformation=&applyVCSProjection=false&returnIdsOnly=false&returnUniqueIdsOnly=false&returnCountOnly=false&returnExtentOnly=false&returnQueryGeometry=false&returnDistinctValues=false&cacheHint=false&orderByFields=&groupByFieldsForStatistics=&outStatistics=&having=&resultOffset=&resultRecordCount=&returnZ=false&returnM=false&returnExceededLimitFeatures=true&quantizationParameters=&sqlFormat=none&f=pjson&token=). Thanks!

**Comment by orestis-z** [comment]
Is the API still maintained?

**Comment by wobsoriano** [comment]
> Is the API still maintained?

Nope, looks dead. I am using a csv based daily data https://github.com/bumbeishvili/covid19-daily-data

Using it here https://github.com/wobsoriano/covid3d
